src/vm/analysis/trait_checker/tests.rs

```diff
@@ -0,0 +1,490 @@
+    }).unwrap_err();
+    match err.err {
+        CheckErrors::TypeError(TypeSignature::TraitReferenceType(_), TypeSignature::TraitReferenceType(_)) => {},
+        _ => {
+            println!("{:?}", err);
+            panic!("Attempt to call init-factorial should fail!")
+        }
+    }
+}
+
+#[test]
+fn test_dynamic_dispatch_mismatched_args() {
+    let dispatching_contract_src =
+        "(define-trait trait-1 (
+            (get-1 (int) (response uint uint))))
+        (define-public (wrapped-get-1 (contract <trait-1>)) 
+            (contract-call? contract get-1 u0))";
+    let target_contract_src =
+        "(impl-trait .dispatching-contract.trait-1)
+        (define-public (get-1 (x uint)) (ok u1))";
+
+    let dispatching_contract_id = QualifiedContractIdentifier::local("dispatching-contract").unwrap();
+    let target_contract_id = QualifiedContractIdentifier::local("target-contract").unwrap();
+
+    let mut dispatching_contract = parse(&dispatching_contract_id, dispatching_contract_src).unwrap();
+    let mut target_contract = parse(&target_contract_id, target_contract_src).unwrap();
+    let mut marf = MemoryBackingStore::new();
+    let mut db = marf.as_analysis_db();
+
+    let err = db.execute(|db| {
+        type_check(&dispatching_contract_id, &mut dispatching_contract, db, true)?;
+        type_check(&target_contract_id, &mut target_contract, db, true)
+    }).unwrap_err();
+    match err.err {
+        CheckErrors::TypeError(_, _) => {},
+        _ => {
+            println!("{:?}", err);
+            panic!("Attempt to call init-factorial should fail!")
+        }
+    }
+}
+
+#[test]
+fn test_dynamic_dispatch_mismatched_returns() {
+    let dispatching_contract_src =
+        "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))))
+        (define-public (wrapped-get-1 (contract <trait-1>)) 
+            (contract-call? contract get-1 u0))";
+    let target_contract_src =
+        "(impl-trait .dispatching-contract.trait-1)
+        (define-public (get-1 (x uint)) (ok \"buffer\"))";
+
+    let dispatching_contract_id = QualifiedContractIdentifier::local("dispatching-contract").unwrap();
+    let target_contract_id = QualifiedContractIdentifier::local("target-contract").unwrap();
+
+    let mut dispatching_contract = parse(&dispatching_contract_id, dispatching_contract_src).unwrap();
+    let mut target_contract = parse(&target_contract_id, target_contract_src).unwrap();
+    let mut marf = MemoryBackingStore::new();
+    let mut db = marf.as_analysis_db();
+
+    let err = db.execute(|db| {
+        type_check(&dispatching_contract_id, &mut dispatching_contract, db, true)?;
+        type_check(&target_contract_id, &mut target_contract, db, true)
+    }).unwrap_err();
+    match err.err {
+        CheckErrors::BadTraitImplementation(_, _) => {},
+        _ => {
+            println!("{:?}", err);
+            panic!("Attempt to call init-factorial should fail!")
+        }
+    }
+}
+
```

Comment:
Can you add some tests for defining, implementing, and calling functions of traits that have multiple functions defined?

Reply:
In addition, can you add tests to verify that a contract that does not define all trait methods will not type-check?

Reply:
Re-upping this comment above :point_up: